File: packages/sdk/src/agent-tool.ts
import { DigitalCard } from './client.js';
import { InvalidResolveInputError } from './errors.js';
import type { DigitalCardClientOptions, ResolveResult } from './types.js';

export type EILResolveToolInput = {
  domain?: string;
  handle?: string;
};

export {
  buildEILResolveToolDefinition,
  buildEILResolveToolDefinitions,
  buildResolveInputJsonSchema,
  exportToolDefinitionsJson,
  EIL_RESOLVE_TOOL_NAME,
  EIL_RESOLVE_TOOL_DESCRIPTION,
} from './tool-schema-export.js';

export type {
  AnthropicToolDefinition,
  EILResolveToolDefinitions,
  GeminiFunctionDeclaration,
  JsonSchemaObject,
  OpenAIToolDefinition,
} from './tool-schema-export.js';

/** Invoke handler for tool-calling agents (framework-agnostic) */
export async function invokeEILResolve(
  input: EILResolveToolInput,
  options?: DigitalCardClientOptions
): Promise<ResolveResult> {
  if (input.domain) {
    return DigitalCard.resolve({ domain: input.domain }, options);
  }
  if (input.handle) {
    return DigitalCard.resolve({ handle: input.handle }, options);
  }
  throw new InvalidResolveInputError();
}

File: packages/python/eil_card/agent_tool.py
from __future__ import annotations

import json
from typing import Any

from eil_card.client import DigitalCard, DigitalCardClient
from eil_card.errors import InvalidResolveInputError
from eil_card.tool_schema import EIL_RESOLVE_TOOL_DESCRIPTION, EIL_RESOLVE_TOOL_NAME


async def invoke_eil_resolve_async(
    *,
    domain: str | None = None,
    handle: str | None = None,
    registry_base_url: str = "https://eilcard.com",
    api_key: str | None = None,
    timeout: float = 10.0,
    skip_well_known_fallback: bool = False,
) -> dict[str, Any]:
    """Async wrapper — runs sync resolve in thread pool if needed by frameworks."""
    return invoke_eil_resolve(
        domain=domain,
        handle=handle,
        registry_base_url=registry_base_url,
        api_key=api_key,
        timeout=timeout,
        skip_well_known_fallback=skip_well_known_fallback,
    )


def invoke_eil_resolve(
    *,
    domain: str | None = None,
    handle: str | None = None,
    registry_base_url: str = "https://eilcard.com",
    api_key: str | None = None,
    timeout: float = 10.0,
    skip_well_known_fallback: bool = False,
) -> dict[str, Any]:
    """Framework-agnostic handler for tool-calling agents."""
    client = DigitalCardClient(
        registry_base_url=registry_base_url,
        api_key=api_key,
        timeout=timeout,
        skip_well_known_fallback=skip_well_known_fallback,
    )
    if domain:
        return client.resolve(domain=domain)
    if handle:
        return client.resolve(handle=handle)
    raise InvalidResolveInputError()


def format_resolve_tool_result(result: dict[str, Any]) -> str:
    return json.dumps(result, indent=2, ensure_ascii=False)


def format_resolve_tool_error(entity: str, error: Exception | str) -> str:
    message = str(error)
    return f"Could not resolve EIL identity for {entity}: {message}"
